examples/fluids/navierstokes.c

```diff
@@ -1165,20 +1165,50 @@ int main(int argc, char **argv) {
-      // Face 4 is inflow and Face 2 is outflow
-      bc.ninflow = bc.noutflow = 1;
-      bc.inflow[0] = 4; bc.outflow[0] = 2;
-      // Face 3 is slip
-      bc.nslip[0] = bc.nslip[2] = 0; bc.nslip[1] = 1;
-      bc.slips[1][0] = 3;
-      // Face 1 is wall
-      bc.nwall = 1;
-      bc.walls[0] = 1;
-      break;
+      bc.nwall = bc.nslip[0] = bc.nslip[1] = bc.nslip[2] = 0; // No wall BCs, and slip BCs will be determined according to the wind vector.
+      if (wind[0] == 0 && wind[1] == 0) {
+          SETERRQ2(PETSC_COMM_SELF, PETSC_ERR_ARG_INCOMP,
+                   "No translation with problem_advection_wind_translation %f,%f. At least one of the elements needs to be non-zero.",
+                   wind[0], wind[1]);
+        break;} else if (wind[0] == 0 && wind[1] > 0) {
+          bc.ninflow = bc.noutflow = 1; // Face 1 is inflow and Face 3 is outflow
+          bc.inflow[0] = 1; bc.outflow[0] = 3;
+          bc.nslip[0] = 2;              // Faces 2 and 4 are slip
+          bc.slips[0][0] = 2; bc.slips[0][1] = 4;
+        break;} else if (wind[0] == 0 && wind[1] < 0) {
+          bc.ninflow = bc.noutflow = 1; // Face 3 is inflow and Face 1 is outflow
+          bc.inflow[0] = 3; bc.outflow[0] = 1;
+          bc.nslip[0] = 2;              // Faces 2 and 4 are slip
+          bc.slips[0][0] = 2; bc.slips[0][1] = 4;
+        break;} else if (wind[0] > 0 && wind[1] == 0) {
+          bc.ninflow = bc.noutflow = 1; // Face 4 is inflow and Face 2 is outflow
+          bc.inflow[0] = 4; bc.outflow[0] = 2;
+          bc.nslip[1] = 2;              // Faces 1 and 3 are slip
+          bc.slips[1][0] = 1; bc.slips[1][1] = 3;
+        break;} else if (wind[0] < 0 && wind[1] == 0) {
+          bc.ninflow = bc.noutflow = 1; // Face 2 is inflow and Face 4 is outflow
+          bc.inflow[0] = 2; bc.outflow[0] = 4;
+          bc.nslip[1] = 2;              // Faces 1 and 3 are slip
+          bc.slips[1][0] = 1; bc.slips[1][1] = 3;
+        break;} else if (wind[0] > 0 && wind[1] > 0) {
+          bc.ninflow = bc.noutflow = 2; // Faces 1 and 4 are inflow and Faces 2 and 3 are outflow
+          bc.inflow[0]  = 1; bc.inflow[1]  = 4;
+          bc.outflow[0] = 2; bc.outflow[1] = 3;
+        break;} else if (wind[0] > 0 && wind[1] < 0) {
+          bc.ninflow = bc.noutflow = 2; // Faces 3 and 4 are inflow and Faces 1 and 2 are outflow
+          bc.inflow[0]  = 3; bc.inflow[1]  = 4;
+          bc.outflow[0] = 1; bc.outflow[1] = 2;
+        break;} else if (wind[0] < 0 && wind[1] > 0) {
+          bc.ninflow = bc.noutflow = 2; // Faces 1 and 2 are inflow and Faces 3 and 4 are outflow
+          bc.inflow[0]  = 1; bc.inflow[1]  = 2;
+          bc.outflow[0] = 3; bc.outflow[1] = 4;
+        break;} else if (wind[0] < 0 && wind[1] < 0) {
+          bc.ninflow = bc.noutflow = 2; // Faces 2 and 3 are inflow and Faces 1 and 4 are outflow
+          bc.inflow[0]  = 2; bc.inflow[1]  = 3;
+          bc.outflow[0] = 1; bc.outflow[1] = 4;
+        break;}
```

Comment:
Yeah, it could be even cumbersome to implement it on our 3d problems. I thought we were only concerned about this mesh. But, of course, considering a more general case is wise.

we could keep only one QFunction for both `inflow` and `outflow` and after computing `sign(dot(velocity, normal))`, let a `switch-case` decide which operation is required. My only concern is how to update  `SetUpDM` when `dot(velocity, normal) = 0`. Is it possible to update a struct inside the QFunction?

Reply:
For advection, it doesn't matter because flux is multiplied by velocity, so if the velocity is zero, the flux (integrand) is zero. For Euler/NS, we have only the incoming/outgoing acoustic wave, but we normally identify inflow vs outflow anyway. Zero velocity can happen in low-speed recirculating flows, but we'd have to think through the details and the viscous contribution is likely significant.

So yes, I would go ahead with a single in/outflow QFunction.

Reply:
@jedbrown , could you please take a look at my latest changes? Is this what you meant? I think I am missing something because I can't reproduce the correct behavior of the blob with `translation` mode. I have been trying to find the bug for hours now but I can't see what goes wrong.